### ISSUE TITLE
Give warning for likely mistake in slice assignment.

### ISSUE THREAD

**Issue by JonathanHallstrom** [ziglang/zig#17469]
### Zig Version

0.12.0-dev.817+54e7f58fc

### Steps to Reproduce and Observed Output

1. Compile the code:
```zig
        const a = [_]u1 {0};
        var slice = a[0..];
        slice = slice[0..];
```
Currently there is no warning.

### Expected Output

Since this is likely a mistake because the assignment `slice = slice[0..];` is a no-op, I would expect there to be some form of warning.

**Comment by tau-dev** [comment]
That warning could easily false positive, because generic code may well want to rely on the fact that `[0..]` can be applied to an array (taking a reference to it) as well as a slice (on which it is an actual no-op). The policy of the Zig compiler has always been to emit a diagnostic like that only if there is a different way of expressing the thing it to make clear that you did actually mean the thing (e.g. unused local -> explicit discard; ignoring an error -> `catch unreachable`).

Now slicing from 0 instead of some other number really doesn't sound so error-prone an operation that we'd need to cover it with a special "I really mean it" expression, especially given that we still couldn't cover any other typo in the slice's indices.

**Comment by nektro** [comment]
i support this since arrays can still be converted to slices with `&`
note for OP that in zig there's only errors, no warnings

**Comment by tau-dev** [comment]
@nektro yeah, but now some generic code would have to special-case arrays vs. slices, for very little practical gain

**Comment by nektro** [comment]
i dont think it would be that big of an issue in practice. but the best way to find out find out would be to implement the error and see how much breaks. generally for things like this i dont think we should let hypothetical generic code stop us from introducing new errors unless its a common idiom.

**Comment by tau-dev** [comment]
Fair enough, I can't point to any examples. It's just that the potential mistake to be prevented here seems entirely hypothetical too. Sure, typos happen, but they make up a small fraction of bugs (in my experience, at least), and trying to cover this incredibly particular part of that fraction does not seem worth the effort

**Comment by tau-dev** [comment]
Okay, different angle than "index typo": searching the standard library for `[0..]`, it looks like the majority of its uses are completely redundant. I may still not like it, but there is definitely precedent in the compiler for disallowing redundancy, e.g. of the `comptime` operator.

**Comment by andrewrk** [comment]
zig has no warnings
